### PR TITLE
feat: add Browser automation CLI via DOMShell MCP server

### LINKED ISSUE #90: [Feature] Browser automation via Accessibility Tree CLI
Great project — the thesis that CLI wrappers make software agent-native maps exactly to what I've been building for browser automation.
I noticed browsers aren't in the current supported applications list. I've been working on this problem with [DOMShell](https://github.com/apireno/DOMShell), which maps Chrome's Accessibility Tree to a virtual filesystem. Agents navigate with ls, cd, grep, click — same shell idioms you're using for desktop apps. In benchmarks with Claude, it cut API calls by 50% vs screenshot-based browsing.
The Accessibility Tree is the key insight for browsers — it's what Chrome already computes for screen readers, so it gives you a stable, semantic interface without brittle DOM scraping or CSS selectors. It maps naturally to a directory hierarchy: page → sections → elements.
A few questions:

Is browser support on your roadmap?
Would a Chrome/browser harness fit your CLI generation pipeline, or is the architecture specific to desktop apps with programmatic APIs?
Would you be open to a collaboration or integration? DOMShell currently runs as an MCP server — it could potentially be adapted as a CLI-Anything harness.

Happy to discuss further. The "CLI for everything" vision is exactly right — browsers are just the next frontier.

### PR DESCRIPTION
**Pull request by furkankoykiran**
## Description

This PR implements browser automation support for CLI-Anything by integrating DOMShell's MCP server. DOMShell maps Chrome's Accessibility Tree to a virtual filesystem, enabling filesystem-first browser automation with `ls`, `cd`, `cat`, `grep`, and `click` commands.

Resolves #90

## Type of Change

- [x] **New Software CLI** — adds a CLI harness for a new application

---

### For New Software CLIs

- [x] `<SOFTWARE>.md` SOP document exists at `browser/agent-harness/HARNESS.md`
- [x] `SKILL.md` exists inside the Python package (`cli_anything/browser/skills/SKILL.md`)
- [x] Unit tests at `cli_anything/browser/tests/test_core.py` are present and pass without backend (31/31 passed)
- [x] E2E tests at `cli_anything/browser/tests/test_full_e2e.py` are present (10 tests, skip without DOMShell)
- [x] `README.md` includes the new software (with link to harness directory)
- [x] `registry.json` includes an entry for the new software
- [x] `repl_skin.py` in `utils/` is an unmodified copy from the plugin

### General Checklist

- [x] Code follows existing patterns and conventions
- [x] `--json` flag is supported on any new commands
- [x] Commit messages follow the conventional format (`feat:`)
- [x] I have tested my changes locally

### Unit Tests (31/31 PASSED)
```
TestSession:        11 tests ✅
TestPageModule:      6 tests ✅
TestFsModule:       13 tests ✅
TestDaemonMode:      2 tests ✅
```

### E2E Tests (10 skipped - DOMShell not installed)
Tests correctly use `pytest.skipif` when dependencies are unavailable.

## Key Features

- **MCP Backend Pattern**: First CLI-Anything harness to use an MCP server as a backend
- **Daemon Mode**: Optional persistent connection for faster interactive use
- **Page State Management**: URL, working directory, navigation history tracking
- **Command Groups**:
  - `page`: open, reload, back, forward, info
  - `fs`: ls, cd, cat, grep, pwd (filesystem primitives)
  - `act`: click, type (element actions)
  - `session`: status, daemon-start, daemon-stop

## Architecture

```
┌─────────────┐     ┌──────────────────┐     ┌─────────────────────┐
│   CLI Cmd   │────▶│ browser_cli.py   │────▶│  MCP Backend       │
│  (Click)    │     │  (command groups)│     │ (domshell_backend) │
└─────────────┘     └──────────────────┘     └─────────┬───────────┘
                                                           │
                    ┌─────────────────────────────────┼────────────────┐
                    │                                 │                │
                    ▼                                 ▼                ▼
            ┌───────────────┐                 ┌────────────────┐    ┌─────────────┐
            │ Spawn npx     │                 │  DOMShell MCP  │    │   Chrome    │
            │ subprocess    │◀──stdio─────────▶│   server      │◀───│ + Ext       │
            └───────────────┘                 └────────────────┘    └─────────────┘
```

## Documentation

- `browser/agent-harness/HARNESS.md` - Browser-specific SOP
- `cli-anything-plugin/HARNESS.md` - Updated with MCP Backend Pattern section
- `cli_anything/browser/README.md` - Installation and usage guide
- `cli_anything/browser/skills/SKILL.md` - Agent skill definition

## Usage Example

```bash
# Open a page
cli-anything-browser page open https://example.com

# Explore structure
cli-anything-browser fs ls /
cli-anything-browser fs cd /main
cli-anything-browser fs grep "Login"

# Click element
cli-anything-browser act click /main/button[0]

# JSON output
cli-anything-browser --json fs ls /

# REPL mode
cli-anything-browser
```

## Dependencies

- Python 3.10+
- Node.js and npx (for DOMShell MCP server)
- Chrome/Chromium with DOMShell extension
- `mcp>=0.1.0` (MCP Python SDK)

### CODE REVIEW

File: browser/agent-harness/cli_anything/browser/browser_cli.py [hunk 1]
```diff
@@ -0,0 +1,422 @@
...
+        if _json_output:
+            click.echo(json.dumps({"error": msg, "type": "dependency_error"}))
+        else:
+            click.echo(f"Error: {msg}", err=True)
+            click.echo(
+                "\nInstall DOMShell Chrome extension:\n"
+                "  https://chromewebstore.google.com/detail/domshell"
+            )
+        sys.exit(1)
+
```
[the remark this answers is not in the document]
Comment (furkankoykiran) [reply browser/agent-harness/cli_anything/browser/browser_cli.py]: Fixed in commit 99ff743. The availability check is now skipped for `--help` and `--version` flags, allowing help to work without DOMShell installed. This aligns with other CLI-Anything harnesses.

File: browser/agent-harness/cli_anything/browser/README.md [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+# Start daemon
+cli-anything-browser session daemon-start
+
+# Run commands (uses persistent connection)
+cli-anything-browser fs ls /
+cli-anything-browser fs cd /main
+
+# Stop daemon when done
+cli-anything-browser session daemon-stop
+```
```
[the remark this answers is not in the document]
Comment (furkankoykiran) [reply browser/agent-harness/cli_anything/browser/README.md:121]: Fixed in commit b88b545. Updated documentation to clarify that daemon mode only works within a single running process (REPL or `--daemon` flag). State does not persist across separate CLI invocations.

File: browser/agent-harness/cli_anything/browser/skills/SKILL.md [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+
+Check `is_available()` return value before running commands.
+
+### Daemon Mode for Efficiency
+
+For agent workflows with multiple commands, use daemon mode:
+
+1. Start daemon: `cli-anything-browser session daemon-start`
+2. Run commands: Each command reuses the MCP connection
+3. Stop daemon: `cli-autocomplete-browser session daemon-stop`
```
[the remark this answers is not in the document]
Comment (furkankoykiran) [reply browser/agent-harness/cli_anything/browser/skills/SKILL.md]: Fixed in commit a41c5be. Corrected command name from `cli-autocomplete-browser` to `cli-anything-browser`.

File: cli-anything-plugin/HARNESS.md [hunk 4]
```diff
@@ -644,6 +701,7 @@ This same SOP applies to any GUI application:
 | Shotcut/Kdenlive | `melt` or `ffmpeg` | .mlt (XML) | `apt install melt ffmpeg` | Build MLT XML → melt/ffmpeg renders video |
 | Audacity | `sox` | .aup3 | `apt install sox` | Generate sox commands → sox processes audio |
 | OBS Studio | `obs-websocket` | scene.json | `apt install obs-studio` | WebSocket API → OBS captures/records |
+| Browser (DOMShell) | `npx @apireno/domshell` (MCP) | Accessibility Tree (virtual FS) | `npx install -g npx` + Chrome ext | MCP SDK → DOMShell tools → filesystem navigation |
```
[the remark this answers is not in the document]
Comment (furkankoykiran) [reply cli-anything-plugin/HARNESS.md]: Fixed in commit 6129577. Corrected installation command from `npx install -g npx` to `npm install -g npx`.

File: browser/agent-harness/cli_anything/browser/tests/test_full_e2e.py [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+import pytest
+from click.testing import CliRunner
+
+from cli_anything.browser.utils.domshell_backend import is_available
+from cli_anything.browser.browser_cli import cli
+
+# Skip all tests if DOMShell is not available
+pytestmark = pytest.mark.skipif(
+    not is_available()[0],
+    reason="DOMShell MCP server not available. Install from Chrome Web Store."
```
[the remark this answers is not in the document]
Comment (furkankoykiran) [reply browser/agent-harness/cli_anything/browser/tests/test_full_e2e.py]: Fixed in commit a8fd4d6. E2E tests now use `DOMSHELL_E2E` environment variable for gating, avoiding unwanted `npx` invocation during test collection. Tests only run when explicitly enabled via `DOMSHELL_E2E=1`.

File: browser/agent-harness/cli_anything/browser/core/page.py [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+        >>> go_back(session)
+        {"url": "https://previous.com", "status": "navigated"}
+    """
+    previous_url = session.go_back()
+    if previous_url is None:
+        return {"error": "No back history available"}
+
+    use_daemon = session.daemon_mode
+    result = backend.open_url(previous_url, use_daemon=use_daemon)
+    session.set_url(previous_url, record_history=False)
```
[the remark this answers is not in the document]
Comment (furkankoykiran) [reply browser/agent-harness/cli_anything/browser/core/page.py]: Fixed in commit 1dfce31. `go_back()` now directly calls `backend.back()` MCP tool instead of reconstructing navigation via `open_url()`. Session state is updated based on the backend result.

File: browser/agent-harness/cli_anything/browser/core/page.py [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+    next_url = session.go_forward()
+    if next_url is None:
+        return {"error": "No forward history available"}
+
+    use_daemon = session.daemon_mode
+    result = backend.open_url(next_url, use_daemon=use_daemon)
+    session.set_url(next_url, record_history=False)
+    return result
+
+
```
[the remark this answers is not in the document]
Comment (furkankoykiran) [reply browser/agent-harness/cli_anything/browser/core/page.py]: Fixed in commit 1dfce31. `go_forward()` now directly calls `backend.forward()` MCP tool instead of reconstructing navigation via `open_url()`. Session state is updated based on the backend result.

File: browser/agent-harness/cli_anything/browser/utils/domshell_backend.py [hunk 8]
```diff
@@ -0,0 +1,386 @@
...
+    if _daemon_session is None:
+        return
+
+    try:
+        await _daemon_session.__aexit__(None, None, None)
+        if _daemon_read and _daemon_write:
+            await stdio_client(
+                StdioServerParameters(command=DEFAULT_SERVER_CMD, args=DEFAULT_SERVER_ARGS)
+            ).__aexit__(None, None, None)
+    except Exception:
```
[the remark this answers is not in the document]
Comment (furkankoykiran) [reply browser/agent-harness/cli_anything/browser/utils/domshell_backend.py:190]: Fixed in commit 017a6ce. Stored the `stdio_client` context manager globally as `_daemon_client_context` to ensure proper cleanup. Now `_stop_daemon()` calls `__aexit__` on the same instance that was entered.

File: browser/agent-harness/cli_anything/browser/browser_cli.py [hunk 9]
```diff
@@ -0,0 +1,422 @@
...
+                skin.print_goodbye()
+                break
+            if line.lower() == "help":
+                skin.help(_repl_commands)
+                continue
+
+            # Parse and execute command
+            args = line.split()
+            try:
+                cli.main(args, standalone_mode=False)
```
[the remark this answers is not in the document]
Comment (furkankoykiran) [reply browser/agent-harness/cli_anything/browser/browser_cli.py]: Fixed in commit 99ff743. REPL now uses `shlex.split()` to preserve quoted arguments, with a fallback to `line.split()` for unbalanced quotes. This allows commands like `act type /path "Hello, World!"` to work correctly.

File: browser/agent-harness/cli_anything/browser/README.md [hunk 10]
```diff
@@ -0,0 +1,263 @@
...
+cli-anything-browser page open https://example.com
+
+# List elements at root
+cli-anything-browser fs ls /
+
+# Navigate into a section
+cli-anything-browser fs cd /main
+
+# List elements in current directory
+cli-anything-browser fs ls
```
[the remark this answers is not in the document]
Comment (furkankoykiran) [reply browser/agent-harness/cli_anything/browser/README.md:61]: Fixed in commit b88b545. Added note clarifying that one-shot commands each start with a fresh session (no URL or working directory). For stateful workflows like `cd` followed by pathless `ls`, the REPL should be used instead.

**Comment by omerarslan0** [comment]
PR'ı düzenlemeyi bitirince raporunu bekliyorum.


Once you've finished editing the PR, I'll be waiting for your report.

**Comment by furkankoykiran** [comment]
All 10 Copilot review points are addressed. Here's the summary:

The CLI now works without DOMShell when you run `--help` or `--version` (commit 99ff743). REPL commands with quoted arguments like `act type /path "Hello, World!"` now parse correctly thanks to `shlex.split()`.

For navigation, `go_back()` and `go_forward()` call the backend MCP tools directly instead of trying to reconstruct history (1dfce31). Daemon mode also got a lifecycle fix—it now properly stores the stdio_client context so cleanup actually works (017a6ce).

On the testing side, E2E tests won't accidentally trigger `npx` during collection anymore. They're gated by `DOMSHELL_E2E=1` (a8fd4d6).

Documentation updates: corrected the command name typo (a41c5be), fixed the `npx` install instruction (6129577), and clarified that daemon mode and state persistence only work within a single running process, not across separate CLI invocations (b88b545).

31/31 unit tests pass. E2E tests skip cleanly when DOMShell isn't available.

The `feat/browser-domshell-mcp-harness` branch is ready for merge.

Her şey hazır, merge edebilirsiniz :)

**Comment by omerarslan0** [comment]
### English

Great contribution — this is a well-structured harness and a meaningful milestone as the first MCP-backend pattern in CLI-Anything. The filesystem-first approach via DOMShell is a smart design choice. However, there are a few issues that should be addressed before merge:

**Critical**

1. **Daemon mode context manager leak** (`domshell_backend.py`): `_start_daemon()` calls `__aenter__` on `stdio_client(...)` but never stores the context manager reference. Then `_stop_daemon()` creates a *brand new* `stdio_client` instance and calls `__aexit__` on it — so the original subprocess is never cleaned up. Consider using `AsyncExitStack` or storing the CM reference in a module global.

2. **`go_back()` / `go_forward()` rely on in-memory history instead of MCP tools** (`page.py`): These functions use `Session`'s local history stack and re-navigate via `backend.open_url()`. This means `page back` is completely broken in one-shot CLI usage (history is always empty) and can drift from the browser's actual history. DOMShell already provides `domshell_back` and `domshell_forward` tools — please use those instead.

3. **One-shot state doesn't persist** — The README and SKILL.md show workflows like `fs cd /main` followed by `fs ls` as separate CLI invocations, but `Session` is purely in-memory. These examples silently fail. Either add file-based session persistence, or clearly document these as REPL-only workflows.

**Medium**

4. **`--help` fails without DOMShell installed** — The availability check + `sys.exit(1)` runs in the top-level `cli()` callback, so `cli-anything-browser --help` crashes if DOMShell isn't present. Other harnesses let help work without dependencies. Move the check into commands that actually call the backend.

5. **REPL breaks on quoted arguments** — `line.split()` at line 398 doesn't handle quotes. `act type /path "Hello, World!"` splits incorrectly. Use `shlex.split(line)` with a fallback.

6. **E2E skip triggers `npx` during test collection** — `is_available()` runs `npx @apireno/domshell --version` at import time, which can trigger a package download in CI. An env-var gate (`DOMSHELL_E2E=1`) would be safer.

**Minor**

7. **Typo in SKILL.md**: `cli-autocomplete-browser session daemon-stop` → `cli-anything-browser session daemon-stop`
8. **Invalid install command in HARNESS.md table**: `npx install -g npx` → `npm install -g npx` (or just note that npx ships with Node)

Overall this is a valuable addition to the project. Looking forward to the fixes!

---

### Türkçe

Güzel bir katkı, CLI-Anything'in ilk MCP backend tabanlı harness'ı olması önemli bir adım. DOMShell üzerinden filesystem-first yaklaşım akıllıca tasarlanmış. Merge öncesinde düzeltilmesi gereken birkaç konu var:

**Kritik**

1. **Daemon mode'da context manager sızıntısı** (`domshell_backend.py`): `_start_daemon()` context manager'ı saklamamış, `_stop_daemon()` ise yeni bir `stdio_client` instance'ı üzerinde `__aexit__` çağırıyor — yani orijinal subprocess asla temizlenmiyor. `AsyncExitStack` kullanılabilir.

2. **`go_back()` / `go_forward()` lokal history'ye dayanıyor** (`page.py`): DOMShell'in `domshell_back` ve `domshell_forward` tool'ları varken, local Session stack'i + `open_url` ile navigasyon yapılmış. One-shot kullanımda history her zaman boş olacağından bu komutlar çalışmaz. Backend'in native tool'larını kullan.

3. **One-shot komutlarda state kayboluyor** — README ve SKILL.md'deki örnekler (`fs cd /main` → ayrı komut olarak `fs ls`) aslında çalışmıyor çünkü Session in-memory. Ya dosya tabanlı session persistence ekle, ya da bunları REPL-only olarak belgele.

**Orta**

4. **`--help` DOMShell olmadan çalışmıyor** — Bağımlılık kontrolü üst seviye `cli()` callback'inde. Sadece backend'e erişen komutlara taşınmalı.

5. **REPL'de tırnaklı argümanlar bozuluyor** — `line.split()` yerine `shlex.split(line)` kullanılmalı.

6. **E2E testlerde `npx` test collection sırasında tetikleniyor** — `DOMSHELL_E2E=1` gibi bir env-var gate daha güvenli olur.

**Küçük**

7. **SKILL.md'de typo**: `cli-autocomplete-browser` → `cli-anything-browser`
8. **HARNESS.md'de geçersiz komut**: `npx install -g npx` → `npm install -g npx`

Genel olarak projeye değerli bir katkı, düzeltmelerden sonra merge'e hazır olur. Başarılar!

**Comment by furkankoykiran** [comment]
Thank you for the detailed review! However, I want to clarify that all 8 issues you mentioned have already been addressed in my previous commits. Here's the mapping:

**Already Fixed:**

1. **Daemon context manager leak** — Fixed in commit 017a6ce. The `_daemon_client_context` global now stores the stdio_client context manager, and `_stop_daemon()` calls `__aexit__` on the same instance. See my reply: [comment](https://github.com/HKUDS/CLI-Anything/pull/118#discussion_r2969991593)

2. **go_back()/go_forward() use MCP tools** — Fixed in commit 1dfce31. Both functions now call `backend.back()` and `backend.forward()` directly. See replies: [go_back](https://github.com/HKUDS/CLI-Anything/pull/118#discussion_r2969991436), [go_forward](https://github.com/HKUDS/CLI-Anything/pull/118#discussion_r2969991518)

3. **One-shot state persistence** — Documented in commit b88b545. README now clarifies that one-shot commands start with a fresh session and REPL should be used for stateful workflows. See reply: [comment](https://github.com/HKUDS/CLI-Anything/pull/118#discussion_r2969991703)

4. **--help works without DOMShell** — Fixed in commit 99ff743. Availability check is skipped for `--help` and `--version` flags. See reply: [comment](https://github.com/HKUDS/CLI-Anything/pull/118#discussion_r2969990677)

5. **REPL quoted arguments** — Fixed in commit 99ff743. Now uses `shlex.split()` with fallback for unbalanced quotes. See reply: [comment](https://github.com/HKUDS/CLI-Anything/pull/118#discussion_r2969991643)

6. **E2E env var gating** — Fixed in commit a8fd4d6. Tests now use `DOMSHELL_E2E` environment variable to avoid npx during collection. See reply: [comment](https://github.com/HKUDS/CLI-Anything/pull/118#discussion_r2969991378)

7. **SKILL.md typo** — Fixed in commit a41c5be. Changed `cli-autocomplete-browser` to `cli-anything-browser`. See reply: [comment](https://github.com/HKUDS/CLI-Anything/pull/118#discussion_r2969990925)

8. **npx install command** — Fixed in commit 6129577. Changed to `npm install -g npx`. See reply: [comment](https://github.com/HKUDS/CLI-Anything/pull/118#discussion_r2969991145)

All commits are pushed to the `feat/browser-domshell-mcp-harness` branch. The branch is ready for merge.

Let me know if there's anything else I should address!

**Comment by omerarslan0** [comment]
Sorry for the first review—I missed that the fixes had already been made in subsequent commits. After reviewing the repo and going through all 8 commits one by one, I confirmed that everything has been resolved properly. LGTM, ready to merge!

---

İlk review için özür dilerim — düzeltmelerin sonraki commit'lerde zaten yapıldığını kaçırmışım. Repoyu inceleyip tüm 8 commit'i tek tek inceledikten sonra her şeyin düzgün çözüldüğünü doğruladım. LGTM, merge'e hazır!

**Review by omerarslan0 — DISMISSED** [review]

File: registry.json [hunk 11]
```diff
@@ -61,6 +61,20 @@
...
+      "display_name": "Browser",
+      "version": "1.0.0",
+      "description": "Browser automation via DOMShell MCP server. Maps Chrome's Accessibility Tree to a virtual filesystem for agent-native navigation.",
+      "requires": "Node.js, npx, Chrome + DOMShell extension",
+      "homepage": "https://github.com/apireno/DOMShell",
+      "install_cmd": "pip install git+https://github.com/HKUDS/CLI-Anything.git#subdirectory=browser/agent-harness",
+      "entry_point": "cli-anything-browser",
+      "skill_md": "browser/agent-harness/cli_anything/browser/skills/SKILL.md",
+      "category": "web",
+      "contributor": "CLI-Anything-Team",
```
Comment (yuh-yang) [inline registry.json]: pls put yourself as the contributor

File: cli-anything-plugin/HARNESS.md [hunk 12]
```diff
@@ -424,6 +424,63 @@ the input. Users can't tell anything happened.
 
 **Priority order for rendering:** native engine → translated filtergraph → script.
 
+### MCP Backend Pattern
```
Comment (yuh-yang) [inline cli-anything-plugin/HARNESS.md:427]: It's definitely worthwhile to put MCP backend as a separate example. But it's also important to be cautious about the length of HARNESS.md. Maybe it's time for us to make HARNESS.md progressive disclosure as well, i.e., put examples for different scenarios/use-cases into separate .md files and let the agents retrieve them adaptively.

**Comment by yuh-yang** [comment]
Thanks! I guess this is similar to Browser CLI but it's also valuable to include it in CLI-Anything. I left two comments here. Please put the correct contributor in registry.json. I'll create a PR separately to handle progressive disclosure in HARNESS.md.

**Comment by furkankoykiran** [comment]
Updated the contributor in registry.json to `furkankoykiran` with the correct GitHub URL (commit e23cd94).

Thanks for the review! Looking forward to your separate PR for progressive disclosure in HARNESS.md.

**Comment by furkankoykiran** [comment]
I've addressed all 9 Copilot review comments from March 22:

1. Security (daa6001): `act type` no longer echoes the typed text value, keeping passwords and sensitive input out of terminal scrollback.

2. Documentation (2e99c4f): Added the missing `from typing import Any` import to the HARNESS.md example.

3. Documentation (d8055f4): Clarified the DOMShell error message—npx runs packages rather than installing them.

4. Testing (5f20c59): `test_daemon_start_with_json` now asserts valid JSON instead of silently passing on parse errors.

5. Testing (5f20c59): `test_invalid_path_gives_error` actually checks behavior instead of doing `assert True`.

6. Architecture (d8055f4): Documented the daemon mode event loop limitation as a known v1 constraint.

7. State (no change): Keeping URL-only sync for `go_forward()` per user decision.

8. Error handling (827e5e0): `change_directory()` only updates `working_dir` when the backend succeeds.

9. Performance (daa6001): Cached `is_available()` for the REPL lifetime to avoid repeated npx spawns.

Also fixed test mocks (c341c00) to patch `backend.back()` and `backend.forward()` instead of the old `backend.open_url()` calls.

31 unit tests pass.

@omerarslan0 @yuh-yang - Could you please review when you have a chance?

File: browser/agent-harness/cli_anything/browser/browser_cli.py [hunk 13]
```diff
@@ -0,0 +1,427 @@
...
+@act.command("type")
+@click.argument("path")
+@click.argument("text")
+@handle_error
+def act_type(path, text):
+    """Type text into an input element."""
+    sess = get_session()
+    use_daemon = sess.daemon_mode
+    result = backend.type_text(path, text, use_daemon=use_daemon)
+    output(result, f"Typed '{text}' into: {path}")
```
[the remark this answers is not in the document]
Comment (furkankoykiran) [reply browser/agent-harness/cli_anything/browser/browser_cli.py]: Fixed in commit daa6001. The command now outputs "Typed into: {path}" without revealing the text value, keeping sensitive input like passwords out of terminal scrollback and logs.

File: cli-anything-plugin/HARNESS.md [hunk 14]
```diff
@@ -424,6 +424,63 @@ the input. Users can't tell anything happened.
...
+- You want to integrate AI/agent tools that speak MCP protocol
+
+**Backend wrapper** (`utils/<service>_backend.py`):
+```python
+import asyncio
+from mcp import ClientSession, StdioServerParameters
+from mcp.client.stdio import stdio_client
+
+async def _call_tool(tool_name: str, arguments: dict) -> Any:
+    """Call an MCP tool."""
```
[the remark this answers is not in the document]
Comment (furkankoykiran) [reply cli-anything-plugin/HARNESS.md:447]: Fixed in commit 2e99c4f. Added `from typing import Any` to the code example.

File: browser/agent-harness/cli_anything/browser/tests/test_full_e2e.py [hunk 15]
```diff
@@ -0,0 +1,149 @@
...
+        assert result.exit_code == 0
+        import json
+        try:
+            data = json.loads(result.output)
+            assert "daemon" in data or "error" in data
+        except json.JSONDecodeError:
+            # Fallback: check if error message is present
+            pass
+
+
```
[the remark this answers is not in the document]
Comment (furkankoykiran) [reply browser/agent-harness/cli_anything/browser/tests/test_full_e2e.py]: Fixed in commit 5f20c59. Removed the silent pass and added proper assertion: `assert "daemon" in data or "error" in data`

File: browser/agent-harness/cli_anything/browser/tests/test_full_e2e.py [hunk 16]
```diff
@@ -0,0 +1,149 @@
...
+class TestErrorHandling:
+    """Test error handling for edge cases."""
+
+    def test_invalid_path_gives_error(self, runner):
+        """Invalid path should give clear error."""
+        # Note: This test's behavior depends on DOMShell's error handling
+        result = runner.invoke(cli, ["fs", "cd", "/invalid/path/that/does/not/exist"])
+        # Should not crash, exit code may be 0 or 1 depending on error
+        assert True  # If we get here, no crash
```
[the remark this answers is not in the document]
Comment (furkankoykiran) [reply browser/agent-harness/cli_anything/browser/tests/test_full_e2e.py]: Fixed in commit 5f20c59. Replaced `assert True` with actual assertions: `assert result.exception is None` and `assert result.exit_code in (0, 1)`

File: browser/agent-harness/cli_anything/browser/utils/domshell_backend.py [hunk 17]
```diff
@@ -0,0 +1,389 @@
...
+            False,
+            "npx not found. Install Node.js from https://nodejs.org/ "
+            "Then run: npm install -g npx"
+        )
+
+    if not _check_npx_has_domshell():
+        return (
+            False,
+            "DOMShell not found. Install with: npx @apireno/domshell --version\n"
+            "Note: First run may download the package (10-50 MB)."
```
[the remark this answers is not in the document]
Comment (furkankoykiran) [reply browser/agent-harness/cli_anything/browser/utils/domshell_backend.py]: Fixed in commit d8055f4. Clarified the message to "Run `npx @apireno/domshell --version` once" since npx runs packages rather than installing them.

File: browser/agent-harness/cli_anything/browser/utils/domshell_backend.py [hunk 18]
```diff
@@ -0,0 +1,389 @@
...
+    )
+
+    try:
+        # Store the context manager so we can properly clean it up later
+        _daemon_client_context = stdio_client(server_params)
+        _daemon_read, _daemon_write = await _daemon_client_context.__aenter__()
+        _daemon_session = ClientSession(_daemon_read, _daemon_write)
+        await _daemon_session.__aenter__()
+        await _daemon_session.initialize()
+        return True
```
[the remark this answers is not in the document]
Comment (furkankoykiran) [reply browser/agent-harness/cli_anything/browser/utils/domshell_backend.py:170]: Addressed in commit d8055f4 with a known limitation comment. Documented that asyncio.run() per call creates new event loops—this is a v1 limitation to be addressed in a future PR.

File: browser/agent-harness/cli_anything/browser/core/page.py [hunk 19]
```diff
@@ -0,0 +1,115 @@
...
+    Example:
+        >>> go_forward(session)
+        {"url": "https://next.com", "status": "navigated"}
+    """
+    use_daemon = session.daemon_mode
+    result = backend.forward(use_daemon=use_daemon)
+
+    # Update session state if backend returned a URL
+    if isinstance(result, dict) and "url" in result:
```
[the remark this answers is not in the document]
Comment (furkankoykiran) [reply browser/agent-harness/cli_anything/browser/core/page.py:97]: Per user decision, URL-only sync is sufficient for v1. The Session's history stack is not synced, but the URL is updated from the backend result.

File: browser/agent-harness/cli_anything/browser/core/fs.py [hunk 20]
```diff
@@ -0,0 +1,113 @@
...
+    elif not path.startswith("/"):
+        # Relative path: append to current working_dir
+        if session.working_dir == "/":
+            path = "/" + path
+        else:
+            path = session.working_dir.rstrip("/") + "/" + path
+
+    use_daemon = session.daemon_mode
+    result = backend.cd(path, use_daemon=use_daemon)
+    session.set_working_dir(path)
```
[the remark this answers is not in the document]
Comment (furkankoykiran) [reply browser/agent-harness/cli_anything/browser/core/fs.py]: Fixed in commit 827e5e0. Now only updates working_dir if backend succeeds (no "error" key in result): `if isinstance(result, dict) and "error" not in result:`

File: browser/agent-harness/cli_anything/browser/browser_cli.py [hunk 21]
```diff
@@ -0,0 +1,427 @@
...
+        if not available:
+            if _json_output:
+                click.echo(json.dumps({"error": msg, "type": "dependency_error"}))
+            else:
+                click.echo(f"Error: {msg}", err=True)
+                click.echo(
+                    "\nInstall DOMShell Chrome extension:\n"
+                    "  https://chromewebstore.google.com/detail/domshell"
+                )
+            sys.exit(1)
```
[the remark this answers is not in the document]
Comment (furkankoykiran) [reply browser/agent-harness/cli_anything/browser/browser_cli.py:132]: Fixed in commit daa6001. Added `_availability_cached` module variable to cache the is_available() result for REPL lifetime, avoiding repeated npx subprocess spawns.

**Comment by omerarslan0** [comment]
LGTM! Thx for this.

**Comment by yuh-yang** [comment]
Thanks for this valuable PR!

**Comment by apireno** [comment]
This is a great milestone — thanks @furkankoykiran for building the harness, and @omerarslan0 and @yuh-yang for the thorough review.

I'll pull the PR and test it against DOMShell this week. A few things I'll be looking at:

Tree refresh + daemon mode: DOMShell auto-detects DOM mutations and refreshes the AX tree on the next command. Want to verify the daemon session handles tree invalidation gracefully.
Permission flags: Some commands (click, type, navigate) require --allow-write. I'll confirm the harness passes the right flags on startup.
The cd success-check: Good fix — the AX tree can have paths that look valid but resolve differently after a page mutation, so verifying against the backend response is the right call.
I'll run the unit tests and the E2E suite (DOMSHELL_E2E=1), then smoke-test the REPL with daemon mode against a few real sites.

For a follow-up iteration, I'd suggest adding domshell_eval and domshell_js as advanced commands — they serve as escape hatches when an element doesn't have a clean AX representation. eval is read-only (no flags needed), js requires --allow-write. That separation matters for agent safety.

@yuh-yang on progressive disclosure for HARNESS.md — happy to help structure the MCP backend pattern as a standalone doc. The pattern should be reusable for any MCP server, not just DOMShell.

Will report back after testing.

**Comment by apireno** [comment]
Hey @furkankoykiran —  I tested the browser harness end-to-end against the current DOMShell MCP server and found a few issues.

## What was tested

- **Unit tests (Phase 1):** 31/31 passed ✅
- **E2E tests (Phase 2):** 9/10 passed, 1 failed (port conflict — see Bug 2 below)
- **Manual smoke tests (Phase 3):** Several commands fail due to parameter mismatches

## Already fixed on the DOMShell side

The harness calls `npx @apireno/domshell --version` to check availability, but DOMShell didn't support `--version` — it would try to start the full server and fail if the port was in use. **Fixed in `@apireno/domshell@1.1.2`** (now live on npm). `is_available()` works correctly now.

## 3 remaining issues in the harness

### 1. Port conflict in one-shot mode

Each one-shot command spawns a new `npx @apireno/domshell` subprocess, which tries to bind WebSocket port 9876 and MCP port 3001. If any DOMShell instance is already running (e.g. via Claude Desktop, Cursor, or another MCP client), all commands fail. This also causes the `test_invalid_path_gives_error` E2E test to fail.

Suggestion: default to daemon mode, or detect/reuse an existing DOMShell instance.

### 2. MCP tool parameter name mismatches

The harness sends incorrect parameter names for several tools. Comparing `domshell_backend.py` calls against the actual MCP tool schemas:

| Tool | Harness sends | DOMShell expects |
|---|---|---|
| `domshell_ls` | `{"path": path}` | `{"options": path}` or `{}` |
| `domshell_cat` | `{"path": path}` | `{"name": path}` |
| `domshell_click` | `{"path": path}` | `{"name": path}` |
| `domshell_type` | `{"path": p, "text": t}` | `{"text": t}` (no path param) |
| `domshell_grep` | `{"pattern": p, "path": path}` | `{"pattern": p}` (no path param; has `recursive`, `limit`, `content`) |

`domshell_cd`, `domshell_open`, `domshell_back`, `domshell_forward`, `domshell_reload`, and `domshell_pwd` are correct.

### 3. No `--allow-write` flag for write commands

`DEFAULT_SERVER_ARGS` in `domshell_backend.py` is `["@apireno/domshell"]` — it doesn't include `--allow-write`. This means `act click` and `act type` will always fail with permission errors.

---

Would you prefer I open these as individual issues, or would you like to address them in a follow-up PR on this thread?

**Comment by yuh-yang** [comment]
@apireno 
Thanks for the thorough testing and bug-finding. I do think these 3 issues are worth fixing. Please have a single new PR addressing the issues. Thank you!

**Comment by apireno** [comment]
@yuh-yang PR is up, #135 addresses all 3 issues. Ready for review whenever you get a chance!
